Merge pytest addopts

erjac77 opened the pull request.

erjac77 committed:
Fix PYTEST_ADDOPTS being overriden by the user

# Rust tests and lints will be skipped. Delete if not intended.
[ci skip-rust]

# Building wheels and fs_util will be skipped. Delete if not intended.
[ci skip-build-wheels]

erjac77 committed:
Improve alert handling for duplicate addopts

# Rust tests and lints will be skipped. Delete if not intended.
[ci skip-rust]

# Building wheels and fs_util will be skipped. Delete if not intended.
[ci skip-build-wheels]

jsirois commented on src/python/pants/backend/python/goals/pytest_runner.py:
```diff
@@ -311,14 +313,32 @@ async def setup_pytest_for_target(
+    for popt in pants_add_opts:
+        if any(popt in eopt for eopt in extra_add_opts):
+            logger.warning(
+                f"The option '{popt}' is set by Pants but also in PYTEST_ADDOPTS environment variable. This can cause unexpected behavior."
```
So, I played around with pytest 7.0.1 which is what main currently ships with by default:
https://github.com/erjac77/pants/blob/40e17902092a9b7414473f99bf32e3f3c84e5921/src/python/pants/backend/python/subsystems/pytest.py#L62-L74

It has much more variety in what it accepts. FWICT, there are these variants:
1. `--junit-xml` or `--junitxml`
2. `=` or ` `
3. `-o` or `--override-ini`

For the last, you currently ignore the `-o` prefix to `junit_family` and that's probably fine; so probably this can be ignored. Likewise for 2 you can ignore the introduction of the value separator and you do; so things are still good. For the 1st though, that would need to be checked too.

jsirois commented on src/python/pants/backend/python/goals/pytest_runner.py:
```diff
@@ -311,14 +313,32 @@ async def setup_pytest_for_target(
         ]
+        pants_add_opts.extend(["--cov-report", "--cov-config"])  # --cov not there as multi-allowed.
+
+    # Get extra PYTEST_ADDOPTS env var from each possible sources.
```
This was all starting to feel a bit like too much work / too brittle / etc. so I asked why are we doing things this way / is there another way? 

As to why, I traced history and can find nospecific justification for choosing env var vs directly using args:
+ origin of PYTEST_ADDOPTS: https://github.com/pantsbuild/pants/pull/8901
+ use of PYTEST_ADDOPTS to deal with junit xml: https://github.com/pantsbuild/pants/pull/9594

Here's why pytest has to say about PYTEST_ADDOPTS:
https://docs.pytest.org/en/7.1.x/reference/reference.html?highlight=pytest_addopts#envvar-PYTEST_ADDOPTS

The pre-pend bit made me want to know how multiple instance of a flag works then:
```
python3.9 -mvenv pytest.venv
pytest.venv/bin/pip -q install -U pip
pytest.venv/bin/pip install pytest==7.0.1
touch test.py
$ PYTEST_ADDOPTS="--junit-xml fred.xml" ./pytest.venv/bin/pytest --junit-xml george.xml --junitxml=betty.xml --collect-only -q test.py

------------------------------------------------------------------------------------------------------------------ generated xml file: /home/jsirois/betty.xml ------------------------------------------------------------------------------------------------------------------
no tests collected in 0.00s
```

So it looks like CLI trumps env var and right-most CLI arg trumps leftmost. Since we completely control the CLI args passed to pytest below, it seems we could just insert the args we require there. Our communication channel then becomes fully robust and looking for duplicates in PYTEST_ADDOPTS to warn is just a nicety at that point. In fact, reading about the `-o` option, it turns out its a very difficult nicety to unfailingly present! We only check env vars but we could be stomping a setting in config. We've been doing that all along it turns out! So maybe even the dup warnings checking is going to far, or at least we should only be checking for the `--junitxml` dup, since, FWICT, you cannot specify that option in config.

erjac77 committed:
Check for variant --junit-xml option

# Rust tests and lints will be skipped. Delete if not intended.
[ci skip-rust]

# Building wheels and fs_util will be skipped. Delete if not intended.
[ci skip-build-wheels]
